arcgis-ios-sdk-samples/Route & Directions/Navigate route/NavigateRouteViewController.swift

```diff
@@ -0,0 +1,287 @@
+    ///
+    /// - Parameter result: solved route from the route task.
+    func updateViewpoint(_ result: AGSRouteResult) {
+        // Show the resulting route on the map and save a reference to the route.
+        if let viewPoint = defaultViewPoint {
+            // Reset to initial view point with animation.
+            mapView.setViewpoint(viewPoint, completion: nil)
+        } else if let geometry = result.routes.first?.routeGeometry {
+            mapView.setViewpointGeometry(geometry) { [weak self] _ in
+                // Get the initial zoomed view point.
+                self?.defaultViewPoint = self?.mapView.currentViewpoint(with: .centerAndScale)
+            }
+        }
+    }
+    
+    /// Called in response to the Navigate button being tapped.
+    @IBAction func startNavigation() {
+        navigateButtonItem.isEnabled = false
+        resetButtonItem.isEnabled = true
+        mapView.locationDisplay.autoPanMode = .navigation
+        // If the user navigates the map view away from the location display, activate the recenter button.
+        mapView.locationDisplay.autoPanModeChangedHandler = { [weak self] _ in self?.recenterButtonItem.isEnabled = true }
+        // Start the location data source and location display.
+        mapView.locationDisplay.start()
+    }
+    
+    /// Called in response to the Reset button being tapped.
+    @IBAction func reset() {
+        // Stop the speech, if there is any.
+        speechSynthesizer.stopSpeaking(at: .immediate)
+        // Reset to the starting location for location display.
+        mapView.locationDisplay.dataSource.didUpdate(initialLocation)
+        // Stop the location display as well as datasource generation, if reset before the end is reached.
+        mapView.locationDisplay.stop()
+        mapView.locationDisplay.autoPanMode = .off
+        directionsList.removeAll()
+        statusLabel.text = "Directions are shown here."
+        // Reset the navigation.
+        mapView.locationDisplay.dataSource = makeDataSource(routeResult.routes.first!)
+        routeTracker = makeRouteTracker(routeResult)
+        updateRouteGraphics(remaining: (routeResult.routes.first?.routeGeometry)!)
+        updateViewpoint(routeResult)
+        // Reset buttons state.
+        recenterButtonItem.isEnabled = false
+        resetButtonItem.isEnabled = false
+        navigateButtonItem.isEnabled = true
+    }
+    
+    /// Called in response to the Recenter button being tapped.
+    @IBAction func recenter() {
+        mapView.locationDisplay.autoPanMode = .navigation
+        recenterButtonItem.isEnabled = false
+    }
+    
+    // MARK: UIViewController
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        // Avoid the overlap between the label and the map content.
+        mapView.contentInset.top = CGFloat(statusLabel.numberOfLines) * statusLabel.font.lineHeight
```

Comment:
I didn't notice that the status label was given an explicit number of lines. While not wrong, that property is usually used for labels in a table view cell where you want to be able to assign an arbitrary string to a label and have the label only grow so big. Since that isn't a major concern here, I would expect the number of lines to be zero. As for the content inset, it could be updated dynamically based on the height of the label:

```swift
override func viewDidLayoutSubviews() {
    super.viewDidLayoutSubviews()
    // Avoid the overlap between the label and the map content.
    mapView.contentInset.top = statusLabel.frame.height
}
```

Reply:
![iPhone SE](https://user-images.githubusercontent.com/9660181/78809711-2ca2e500-797c-11ea-81de-a0cf4cce19c6.png)

I set the `numberOfLines` because initially I want to limit the height of the `UILabel`, so that it won't cover too much content on a smaller screen such as iPhone SE. But for this particular example, I also agree that it's OK to leave it as 0.

Reply:
In terms of setting the inset on each turn, I also tried that, but found the UI is a bit clumsy. There's no transition animation between each turn if the inset is update every time. Check the gif below.

![each-turn](https://user-images.githubusercontent.com/9660181/78810980-0120fa00-797e-11ea-855b-7640999ff707.gif)

Reply:
Yeah, that's unfortunately. I guess what you have is fine then.